src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -925,301 +1296,410 @@ internal TransitionRegex<TSet> CreateDerivative()
+            if (_builder._pruneCache.TryGetValue(key, out prunedNode))
             {
-                SymbolicRegexNode<TSet> leaf = head._builder.Epsilon;
-                for (int i = 0; i < continuation.Count; ++i)
-                {
-                    leaf = head._builder.CreateConcat(continuation[i], leaf);
-                }
-                return head._builder.CreateConcat(head, leaf);
+                return prunedNode;
             }
 
-            // Helper function for detecting when an unconditionally nullable pattern is in the transitions and no
-            // more transitions need to be considered. If the derivative simulates backtracking, then in the next
-            // step nothing after an unconditionally nullable state would be considered.
-            // For example, d_a( a|ab ) under backtracking simulation transitions to just epsilon, while without
-            // backtracking simulation it would transition to epsilon|b.
-            // All parts of the function below should consider IsDone and return early when it is true.
-            static bool IsDone(TransitionList transitions, bool simulateBacktracking) =>
-                simulateBacktracking && transitions.Count > 0 && transitions.LastTransition.Node.IsNullable;
-
-            // Nothing and epsilon can't consume a character so they generate no transition
-            if (IsNothing || IsEpsilon)
+            if (!StackHelper.TryEnsureSufficientExecutionStack())
             {
-                return;
+                return StackHelper.CallOnEmptyStack(PruneLowerPriorityThanNullability, context);
             }
 
-            // For both .* and .+ the derivative is .* for any character
-            if (IsAnyStar || IsAnyPlus)
+            switch (_kind)
             {
-                Debug.Assert(!IsDone(transitions, simulateBacktracking));
-                SymbolicRegexNode<TSet> leaf = BuildLeaf(_builder._anyStar, continuation);
-                transitions.Add(leaf, effects?.ToArray() ?? Array.Empty<DerivativeEffect>());
-                // Signal early exit if the leaf is unconditionally nullable
-                return;
+                case SymbolicRegexNodeKind.OrderedOr:
+                    Debug.Assert(_left is not null && _right is not null);
+                    // The left alternative, when nullable, has priority over the right alternative
+                    // Otherwise the left alternative is still active and the right alternative is pruned
+                    prunedNode = _left.IsNullableFor(context) ? _left.PruneLowerPriorityThanNullability(context) :
+                        OrderedOr(_builder, _left, _right.PruneLowerPriorityThanNullability(context), deduplicated: true);
+                    break;
+
+                case SymbolicRegexNodeKind.Concat:
+                    Debug.Assert(_left is not null && _right is not null);
+                    //in a concatenation (X|Y)Z priority is given to XZ when X is nullable (XY is forgotten)
```

Comment:
What does it mean for "XY" to be forgotten?  There is no "XY" in the original expression.

Reply:
Essentially, (X|Y)Z is equivalent to (XZ|YZ) here the branch YZ is forgotten (sorry, XY was a typo, will fix), 
will get beck to addressing the comments after a meeting I have, will go through them all today.

Reply:
Fixed, updated the comment to:
//in a concatenation (X|Y)Z priority is given to XZ when X is nullable
//observe that (X|Y)Z is equivalent to (XZ|YZ) and the branch YZ must be ignored
//when X is nullable (observe that XZ is nullable because this=(X|Y)Z is nullable)